### ISSUE TITLE
SC2016 False positive for variables to full binary path

### ISSUE THREAD

**Issue by TaylorSanchez** [koalaman/shellcheck#970]
I don't know if this is completely a bug or more of a feature request. I don't want to turn off all SC2016 errors, but it pops up a lot in our code because we hardcode the full binary path with a variable. This practice helps keep an 80 line limit, and avoids getting unexpected executables when multiple may exist on a system. This applies to more than just awk, but it is a good example.

#### For bugs
- Rule Id SC2016: 
- My shellcheck version (`shellcheck --version` or "online"):  online
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit

#### For new checks and feature suggestions
- [x] I searched through https://github.com/koalaman/shellcheck/issues and didn't find anything related

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash
# Linux commands and paths
AWK=/usr/bin/awk

slesver=$($AWK '/^VERSION/{print $3}' < /etc/SuSE-release)

echo "${slesver}"

```

#### Here's what shellcheck currently says:
$ shellcheck myscript
 
Line 5:
slesver=$($AWK '/^VERSION/{print $3}' < /etc/SuSE-release)
               ^-- SC2016: Expressions don't expand in single quotes, use double quotes for that.

**Comment by Nightfirecat** [comment]
You can [ignore a specific instance of an error](https://github.com/koalaman/shellcheck/wiki/Ignore#ignoring-one-specific-instance-in-a-file) if needed.

**Comment by TaylorSanchez** [comment]
```sh
 $ shellcheck myscript.sh | grep -c SC2016
49
```
I mean that kind of works, but its a really common issue, I would need to add 49 of those comments for just one of my scripts. That is getting quite messy.

**Comment by kurahaupo** [comment]
I'm not a fan of putting command paths in variables; it violates the
separation of code & data (which is tenuous enough already in the shell,
without making it worse).

They really make a script much harder to read, and make it harder for tools
like Shellcheck to check that what you're doing is sane.

Moreover, they're not exempt from globbing and word-splitting, but most
scripts quietly ignore that (as exemplified by the lack of quoting around
`$AWK` in the original report).

I would simply recommend setting PATH to an appropriate value, but failing
that, using functions like:

    awk() { /usr/local/bin/awk "$@";}

or if you really feel like you MUST use a variable, then at least wrap it so
that it's not cluttering the whole script:

    awk() { "$AWK" "$@"; }

Also note that

    AWK=$(which awk)

is completely pointless, because simply "awk" will necessarily run the same
program as "$AWK".

(It's also suspect because "which" is nonstandard, untrustworthy and slow
compared with "command -v" or "type -p", but I use it here to illustrate the
most common case.)

**Comment by TaylorSanchez** [comment]
The functions seem like an appropriate response. Thanks to your comment I searched around and found some similar discussions on [stackoverflow](https://unix.stackexchange.com/questions/30925/in-bash-when-to-alias-when-to-script-and-when-to-write-a-function).

The goal with the variable $AWK was to avoid unexpected binaries or flags due to binaries in the local directory or aliases. The function handles this idea just as well.

I appreciate you taking the time to give those examples, very helpful.

**Comment by koalaman** [comment]
SC2016 tries to reduce false positives by recognizing common commands where you often single-quote things that look like shell variables (`awk`, `sh`, `perl`, some `sed`) and avoid warning. This is why `echo '$1'` warns but `'awk '$1'` does not. `$AWK` is not on that list.

Naming a function `awk` is a good workaround to get the same logic.

Setting or relying on PATH is arguably better, but that's a different discussion.

**Comment by pkitszel** [comment]
`xargs -r gawk  '$1'` :(

**Comment by brother** [comment]
> ```shell
>  $ shellcheck myscript.sh | grep -c SC2016
> 49
> ```
> 
> I mean that kind of works, but its a really common issue, I would need to add 49 of those comments for just one of my scripts. That is getting quite messy.

An eye for detail is a good eye.
If you stray away from the path given by the default settings it's great to know you actually did check all the occurrences while adding the exceptions.

(or ignore the rule globally - please don't)
